File: src/main/java/com/ject/studytrip/mission/domain/model/Mission.java
package com.ject.studytrip.mission.domain.model;

import static org.springframework.util.StringUtils.hasText;

import com.ject.studytrip.global.common.entity.BaseTimeEntity;
import com.ject.studytrip.stamp.domain.model.Stamp;
import jakarta.persistence.*;
import java.time.LocalDateTime;
import lombok.*;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder(access = AccessLevel.PRIVATE)
public class Mission extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stamp_id", nullable = false)
    private Stamp stamp;

    @Column(nullable = false)
    private String name;

    private boolean completed;

    public static Mission of(Stamp stamp, String name) {
        return Mission.builder().stamp(stamp).name(name).completed(false).build();
    }

    public void updateName(String name) {
        if (hasText(name)) this.name = name;
    }

    public void updateDeletedAt() {
        this.deletedAt = LocalDateTime.now();
    }

    public void updateCompleted() {
        this.completed = true;
    }

    public boolean isCompleted() {
        return completed;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Stamp getStamp() {
        return stamp;
    }
}

File: src/main/kotlin/com/ject/studytrip/mission/domain/model/Mission.kt
package com.ject.studytrip.mission.domain.model

import com.ject.studytrip.global.common.entity.BaseTimeEntity
import com.ject.studytrip.stamp.domain.model.Stamp
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import org.springframework.util.StringUtils.hasText
import java.time.LocalDateTime

@Entity
class Mission protected constructor(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "stamp_id", nullable = false)
    var stamp: Stamp,
    @Column(nullable = false)
    var name: String,
    var completed: Boolean = false,
) : BaseTimeEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set

    companion object {
        fun of(
            stamp: Stamp,
            name: String,
        ): Mission = Mission(stamp, name, false)
    }

    fun updateName(name: String) {
        if (hasText(name)) {
            this.name = name
        }
    }

    fun updateDeletedAt(now: LocalDateTime = LocalDateTime.now()) {
        markDeleted(now)
    }

    fun updateCompleted() {
        this.completed = true
    }

    fun isCompleted(): Boolean = completed
}
